### PR TITLE
Aj/arr stor fix

### PR DESCRIPTION
**Pull request by veriAjinkya**
Array storage test case passed llvm verifier but the generated llvm file did not compile successfully. 
The compilation reported type mismatch.

This PR fixes those types issues.

The issue was due to defaulting to i256* for array. 
Now the type of array are accurate in typeContext and llvm self struct.

### CODE REVIEW

File: lib/libyul2llvm/YulASTVisitor/CodegenVisitor.cpp [hunk 1]
```diff
@@ -425,7 +425,7 @@ llvm::Type *LLVMCodegenVisitor::getLLVMTypeByInfo(llvm::StringRef typeStr,
   if (typeStr.startswith("t_mapping")) {
     return memPtrType;
   } else if (typeStr.find("t_array") != typeStr.npos) {
-    return memPtrType;
+    return intrinsicHelper.getTypeByTypeName(typeStr, STORAGE_ADDR_SPACE);
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/CodegenVisitor.cpp:428]: Does this work for dynamic memory arrays? This seems to assume storage.

File: lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp [hunk 2]
```diff
@@ -52,7 +52,13 @@ llvm::Type *YulIntrinsicHelper::getTypeByTypeName(llvm::StringRef type,
   std::string typeStr = type.str();
   int bitWidth = 0;
   if (type.startswith("t_array")) {
-    return visitor.getDefaultType()->getPointerTo(addrSpaceId);
+    ArrayTypeResult res;
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp:55]:
```suggestion
    ArrayTypeResult res{};
```

File: lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp [hunk 3]
```diff
@@ -52,7 +52,13 @@ llvm::Type *YulIntrinsicHelper::getTypeByTypeName(llvm::StringRef type,
...
+    if (type.startswith("t_array$"))
+      res = patternMatcher.parseArrayTypeFromYul(type);
+    else if (type.startswith("t_array("))
+      res = patternMatcher.parseArrayType(type);
```
Comment (Technius) [inline lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp:59]: The else case is not handled.

File: lib/libyul2llvm/YulASTVisitor/CodegenVisitor.cpp (hunk 1 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/CodegenVisitor.cpp:428]: fixed

File: lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp (hunk 2 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp:55]: done

File: lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp (hunk 3 above)
Comment (veriAjinkya) [reply lib/libyul2llvm/YulASTVisitor/IntrinsicHelper.cpp:59]: added abort to else case

**Review by Technius — APPROVED** [review]

**Pull request merged**